### PR TITLE
[partial refactor] Dedicated attention mask wrapper

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
- add a dedicated class to handle additive attention masks, and centralize some of the typical operations (change types, add to another mask, handle smaller sequences..)
- the sparse machinery is still the same and bools (cc @fmassa), I did not touch that. 
- I think that the above needs to be changed eventually, so that we converge towards additive masks all around, but it means that on the sparse side we need to completely handle a float additive mask, in that a value `a` would both mark the fact that this needs to be computed, and be added a posteriori to the computed attention value (not done right now)
- Fix https://github.com/facebookresearch/xformers/issues/90 (do not slice the attention mask)
- Fix https://github.com/facebookresearch/xformers/issues/91 (no causal flag in the core attention)

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
 author = "Facebook AI Research"
 
 # The full version, including alpha/beta/rc tags
-release = "0.0.5"
+release = "0.0.6"
```
Comment (blefaudeux) [inline docs/source/conf.py:38]: just bumped to mark a difference with the released package (0.0.5)

File: tests/test_pytorch_transformer_parity.py [hunk 2]
```diff
@@ -114,7 +114,6 @@ def test_pytorch_encoder_parity(device=torch.device("cuda")):
             dim_feedforward=4 * EMB,
             dropout=DROP,
             activation=ACTIVATION,
-            layer_norm_eps=1e-05,
```
Comment (blefaudeux) [inline tests/test_pytorch_transformer_parity.py:117]: not present in some pytorch versions, and this is the default anyway

File: xformers/components/attention/__init__.py [hunk 3]
```diff
@@ -77,10 +78,11 @@ class MyAttention(Attention):
 )
 
 
-def maybe_sparsify(matrix):
+def maybe_sparsify(matrix) -> Any:
     # Sparsify if that makes sense
     if torch.count_nonzero(matrix).item() / matrix.numel() > _DENSITY_THRESHOLD:
-        return matrix
+        # If not sparse, then AttentionMask is the reference type
```
Comment (blefaudeux) [inline xformers/components/attention/__init__.py:84]: heads up on that, basically hijacking any non-sparse case to put it under the AttentionMask umbrella

File: xformers/components/attention/attention_mask.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+from typing import Optional, Type, TypeVar
+
+import torch
+
+Self = TypeVar("Self", bound="AttentionMask")
+
+
+class AttentionMask:
```
Comment (blefaudeux) [inline xformers/components/attention/attention_mask.py:14]: tentatively own all attention masks, but this "breaks" (not per say, but not handled here) right now when things are sparsified

File: xformers/components/attention/core.py [hunk 5]
```diff
@@ -69,12 +70,12 @@ def _broadcast_batch(mask, batch_size):
 
 
 def _matmul_with_mask(
-    a: torch.Tensor, b: torch.Tensor, mask: Optional[torch.Tensor]
+    a: torch.Tensor, b: torch.Tensor, mask: Optional[Union[torch.Tensor, "SparseCS"]]
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:73]: that's where the seams break a little in between AttentionMask and the sparse takes. On the bright side, it's all internal. On the dark side, ideally we could have something a little more streamlined overall.

File: xformers/components/attention/core.py [hunk 6]
```diff
@@ -199,25 +201,26 @@ def scaled_query_key_softmax(
...
     # Matmul with mask, if boolean
-    is_bool_mask = att_mask is not None and att_mask.dtype == torch.bool
-    att = _matmul_with_mask(q, k.transpose(-2, -1), att_mask if is_bool_mask else None)
+    if att_mask is not None and isinstance(att_mask, AttentionMask):
+        # Additive mask
+        mask: Optional[Union[SparseCS, torch.Tensor]] = att_mask.values
+    else:
+        mask = att_mask
 
-    # Could also be that the mask was additive
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:205]: @dianaml0 _matmul_with_mask already has an additive mask codepath actually. It was written twice, but the computation was correct (since we only passed the mask to _matmul_with_mask if it was a bool..). Maybe that we should keep the addition here actually and remove it from matmul ?

File: xformers/components/attention/core.py [hunk 7]
```diff
@@ -199,25 +201,26 @@ def scaled_query_key_softmax(
...
-    if att_mask is not None and att_mask.dtype != torch.bool:
-        att = att + att_mask
+    att = _matmul_with_mask(q, k.transpose(-2, -1), mask)
 
     # Softmax to get the attention probabilities
-    att = _softmax(att, causal=causal)
+    is_causal = att_mask.is_causal if isinstance(att_mask, AttentionMask) else False
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:213]: this fixes https://github.com/facebookresearch/xformers/issues/91

File: xformers/components/attention/core.py [hunk 8]
```diff
@@ -231,30 +234,23 @@ def scaled_dot_product_attention(
...
-            if not att_mask.is_sparse:
-                att_mask = att_mask[:seq, :seq]
-            else:
-                logging.warning(
-                    "Mismatching attention mask and sequence length. On the fly correction but this will be slow"
-                )
-                # Loosing sparsity on purpose,
-                # expectation is that moving back and forth dense/sparse will negate the speedup
-                att_mask = att_mask.to_dense().squeeze(0)[:seq, :seq]
+        if isinstance(att_mask, AttentionMask):
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:237]: this is cleaner than before I believe, but the dense/sparse case is exposed

File: xformers/components/attention/global_tokens.py [hunk 9]
```diff
@@ -91,10 +92,24 @@ def forward(
             self.attention_mask = self.attention_mask.to(q.device)
 
         # Mask-aware attention
-        mask = (
-            self.attention_mask if att_mask is None else self.attention_mask & att_mask
+        if att_mask is not None:
```
Comment (blefaudeux) [inline xformers/components/attention/global_tokens.py:95]: same, sparse/dense handling discrepancy, fixing this would be a nice follow up

File: xformers/components/attention/random.py [hunk 10]
```diff
@@ -93,12 +94,24 @@ def forward(
             self.rand_attention_mask = self._get_rand_mask(q.shape).to(q.device)
 
         # Mask-aware attention
-        mask = (
-            self.rand_attention_mask
-            if att_mask is None
-            else self.rand_attention_mask & att_mask
+        if att_mask is not None:
```
Comment (blefaudeux) [inline xformers/components/attention/random.py:97]: same, sparse/dense discrepancy

File: xformers/components/attention/attention_mask.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        self.to_seq_len = additive_mask.shape[0]
+
+    def to_bool(self) -> torch.Tensor:
+        """
+        .. warning: we assume here that True implies that the value should be computed
+        """
+        return self.values != float("-inf")
+
+    @classmethod
+    def from_bool(cls: Type[Self], x: torch.Tensor) -> Self:
```
Comment (dianaml0) [inline xformers/components/attention/attention_mask.py:45]: Really nice to have these!

**Review by dianaml0 — APPROVED** [review]
Really nice to have this all centralized!

File: xformers/components/attention/core.py (hunk 7 above)
Comment (fmassa) [reply xformers/components/attention/core.py:213]: Thanks!

One downside of this approach is that it leaks internal implementation details of `AttentionMask` inside a core function.

My thoughts were to modify instead `_softmax` so that the causal path is just a different function call from a `CausalAttentionMask` Tensor type.

File: xformers/components/attention/core.py (hunk 8 above)
Comment (fmassa) [reply xformers/components/attention/core.py:237]: I still believe we should not be cropping the mask in our core functions. But ok to keep this as is for now

**Review by fmassa — COMMENTED** [review]
Thanks for the work @blefaudeux !

I had some slightly different thoughts about the implementation, but let's move with what you have for now. I think it might be worth getting in a call and discuss.

I propose we get this merged, but IMO we need to re-write this whole file to entirely take into account the things we care about and get the abstractions that best suit our new use-cases.

File: xformers/components/attention/core.py (hunk 7 above)
Comment (blefaudeux) [reply xformers/components/attention/core.py:213]: in that case you pass the mask to softmax, it moves the branching there, why not. Typically softmax do not take masks though, so I think that the interface becomes a little confusing (we'll pass a mask in only to say that it's causal or not)

File: xformers/components/attention/core.py (hunk 8 above)
Comment (blefaudeux) [reply xformers/components/attention/core.py:237]: it's also a matter of deduplicating, else it's something which has to be done in all the corresponding attentions. No strong opinion though, I'm ok to move that up in SDP and others

File: xformers/components/attention/core.py [hunk 12]
```diff
@@ -69,12 +70,12 @@ def _broadcast_batch(mask, batch_size):
...
 
 def _matmul_with_mask(
-    a: torch.Tensor, b: torch.Tensor, mask: Optional[torch.Tensor]
+    a: torch.Tensor, b: torch.Tensor, mask: Optional[Union[torch.Tensor, "SparseCS"]]
 ) -> torch.Tensor:
     if mask is None:
         return a @ b
 
-    if _is_sparse_available:
+    if _is_sparse_available and mask.dtype == torch.bool:
```
[the remark this answers is not in the document]
Comment (blefaudeux) [reply xformers/components/attention/core.py:78]: ideally I think that we should just expose one mask type, it's easier to understand for the user, and do the appropriate thing internally from there (sparse, dense, blocksparse, I agree with @fmassa on that). I think that it's beyond this PR though, so at that time yes if you pass a bool mask it goes the maybe_sparse way, if you pass a float then it's all dense. I can change that, this is a draft !

File: xformers/components/attention/core.py (hunk 8 above)
Comment (blefaudeux) [reply xformers/components/attention/core.py]: fixed that @fmassa :)

File: xformers/components/attention/core.py (hunk 7 above)
Comment (blefaudeux) [reply xformers/components/attention/core.py]: @fmassa is that ok if you propose a change (object oriented if I got it right, the masks handle the dispatch problem) in another PR ?

File: xformers/components/attention/core.py [hunk 13]
```diff
@@ -199,25 +201,26 @@ def scaled_query_key_softmax(
     q = q / math.sqrt(k.size(-1))
 
     # Matmul with mask, if boolean
```
[the remark this answers is not in the document]
Comment (blefaudeux) [reply xformers/components/attention/core.py]: should be fixed

**Comment by blefaudeux** [comment]
landing that as I understood that @dianaml0 and @fmassa were ok with this as a first take, but @fmassa to propose a rewrite where attention mask(s) handle the specific computations redirect (personally ok with that)

File: xformers/components/attention/core.py (hunk 7 above)
Comment (fmassa) [reply xformers/components/attention/core.py]: Sure, I'll send a PR with my proposed changes later this week
